File: haskell/Problem29.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 29
-- Distinct powers

module Problem29 where

import Data.List (nub)

solve :: Int
solve = length . nub $ [a ^ b | a <- range, b <- range] where range = [2..100]

File: c++/src/problem_29.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 29
// Distinct powers

#include <gmpxx.h>

#include <set>

namespace problem_29 {

long solve() {
	std::set<mpz_class> powers;
	for (long i = 2; i <= 100; ++i) {
		for (long j = 2; j <= 100; ++j) {
			mpz_class n;
			mpz_ui_pow_ui(n.get_mpz_t(), i, j);
			powers.insert(n);
		}
	}
	return static_cast<long>(powers.size());
}

} // namespace problem_29
